src/include/common/mpsc_queue.h

```diff
@@ -0,0 +1,106 @@
+#pragma once
+
+// Based on the following design:
+// https://www.1024cores.net/home/lock-free-algorithms/queues/non-intrusive-mpsc-node-based-queue
+
+#include <atomic>
+
+#include "common/assert.h"
+
+namespace kuzu {
+namespace common {
+
+// Producers are completely wait-free.
+template<typename T>
+class MPSCQueue {
+    struct Node {
+        T data;
+        std::atomic<Node*> next;
+
+        explicit Node(T data) : data(std::move(data)), next(nullptr) {}
+    };
+
+    // Head is always present, but always has dummy data. This ensures that it is always easy to
+    // append to the list, without branching in the methods.
+    Node* head;
+    std::atomic<Node*> tail;
+
+    std::atomic<size_t> _approxSize;
+
+public:
+    MPSCQueue() : head(nullptr), tail(nullptr), _approxSize(0) {
+        // Allocate a dummy element.
+        Node* stub = new Node(T());
+        head = stub;
+
+        // Ordering doesn't matter.
+        tail.store(stub, std::memory_order_relaxed);
+    }
+    MPSCQueue(const MPSCQueue& other) = delete;
```

Comment:
hmm Should we use macros in `copy_constructors` instead?

Reply:
Probably. Ill update it later

Reply:
I also rewrote some of that file, since it was a bit unpolished.